File: web/src/domain/measurement.ts
export type Reading = Readonly<{
  systolic: number;
  diastolic: number;
  pulse: number;
}>;

export type MeasurementValues = Reading;

export type FirstReadingCaptured = Readonly<{
  kind: "first-reading-captured";
  reading: Reading;
}>;

export type CalculatedMeasurement = Readonly<{
  kind: "calculated-measurement";
  first: Reading;
  second: Reading;
  result: MeasurementValues;
}>;

export type ConfirmedMeasurement = Readonly<{
  kind: "confirmed-measurement";
  values: MeasurementValues;
}>;

export function captureFirstReading(reading: Reading): FirstReadingCaptured {
  validateReading(reading);
  return { kind: "first-reading-captured", reading: { ...reading } };
}

export function addSecondReading(
  state: FirstReadingCaptured,
  second: Reading,
): CalculatedMeasurement {
  validateReading(second);

  return {
    kind: "calculated-measurement",
    first: state.reading,
    second: { ...second },
    result: {
      systolic: roundHalfUpMean(state.reading.systolic, second.systolic),
      diastolic: roundHalfUpMean(state.reading.diastolic, second.diastolic),
      pulse: roundHalfUpMean(state.reading.pulse, second.pulse),
    },
  };
}

export function confirmMeasurement(state: CalculatedMeasurement): ConfirmedMeasurement {
  return { kind: "confirmed-measurement", values: { ...state.result } };
}

export function roundHalfUpMean(first: number, second: number): number {
  assertPositiveInteger(first, "first");
  assertPositiveInteger(second, "second");

  const lower = Math.min(first, second);
  const difference = Math.max(first, second) - lower;
  return lower + Math.floor(difference / 2) + (difference % 2);
}

function validateReading(reading: Reading): void {
  assertPositiveInteger(reading.systolic, "systolic");
  assertPositiveInteger(reading.diastolic, "diastolic");
  assertPositiveInteger(reading.pulse, "pulse");

  if (reading.diastolic >= reading.systolic) {
    throw new RangeError("diastolic must be lower than systolic");
  }
}

function assertPositiveInteger(value: number, field: string): void {
  if (!Number.isSafeInteger(value) || value <= 0) {
    throw new RangeError(`${field} must be a positive integer`);
  }
}

File: android/domain/src/main/kotlin/com/mitension/domain/Measurement.kt
package com.mitension.domain

/** A temporary reading captured during the two-step input flow. */
data class Reading(
    val systolic: Int,
    val diastolic: Int,
    val pulse: Int,
) {
    init {
        require(systolic > 0) { "systolic must be positive" }
        require(diastolic > 0) { "diastolic must be positive" }
        require(pulse > 0) { "pulse must be positive" }
        require(diastolic < systolic) { "diastolic must be lower than systolic" }
    }
}

/** First flow state. It deliberately exposes no persistible result. */
data class FirstReadingCaptured(val reading: Reading) {
    fun addSecondReading(second: Reading): CalculatedMeasurement =
        CalculatedMeasurement(
            first = reading,
            second = second,
            result = MeasurementValues(
                systolic = roundHalfUpMean(reading.systolic, second.systolic),
                diastolic = roundHalfUpMean(reading.diastolic, second.diastolic),
                pulse = roundHalfUpMean(reading.pulse, second.pulse),
            ),
        )
}

/** Calculated values that still require explicit user confirmation. */
data class CalculatedMeasurement internal constructor(
    val first: Reading,
    val second: Reading,
    val result: MeasurementValues,
) {
    fun confirm(): ConfirmedMeasurement = ConfirmedMeasurement(result)
}

data class MeasurementValues(
    val systolic: Int,
    val diastolic: Int,
    val pulse: Int,
)

/** The only domain type eligible for persistence; it contains no original readings. */
data class ConfirmedMeasurement internal constructor(val values: MeasurementValues)

fun roundHalfUpMean(first: Int, second: Int): Int {
    require(first > 0 && second > 0) { "mean inputs must be positive" }

    val lower = minOf(first, second)
    val difference = maxOf(first, second) - lower
    return lower + difference / 2 + difference % 2
}

